src/wp-includes/media.php

```diff
@@ -1056,6 +1057,11 @@ function wp_get_attachment_image( $attachment_id, $size = 'thumbnail', $icon = f
 
 		$attr = wp_parse_args( $attr, $default_attr );
 
+		// Omit the `decoding` attribute if empty or false.
+		if ( array_key_exists( 'decoding', $attr ) && ! in_array( $attr['decoding'], array( 'async', 'sync', 'auto' ), true ) ) {
```

Comment:
```suggestion
		if ( empty( $attr['decoding'] ) || ! is_string( $attr['decoding'] ) || ! in_array( $attr['decoding'], array( 'async', 'sync', 'auto' ), true ) ) {
```
Requesting confidence check and thoughts:
1. `empty()` values aren't allowed for `decoding`.
2. The current implementation runs `in_array()` - O(n) - for `empty()` values.
3. Replacing `array_key_exists() &&` with `empty( $attr['decoding'] ) ||` prevents this.
4. `unset()` doesn't require that the value is set, so calling it on a non-existent `decoding` key won't cause any issues.
5. An additional `! is_string( $attr['decoding'] ) ||` does the same for non-string values.

Reply:
@costdev so your suggestion here is as an optimization to avoid `in_array` unnecessarily? I guess that seems good, I have a couple of questions:

* How much is this really a gain, how frequently does this get called and how many times would be avoided with the new checks? (ballpark, I don't expect tests) I find the previous version easier to understand immediately)
* Should we also fix the similar check in src/wp-includes/pluggable.php (L:2897)

Reply:
FWIW, I think `empty()` is clearer in intent.

I don't think the `is_string()` check is needed, it should never fail if the developer follows the documented use of the parameter and shouldn't be a problem if an unexpected type is passed. https://3v4l.org/0FNUu

Reply:
@costdev @peterwilsoncc I think the `array_key_exists( 'decoding', $attr )` is appropriate here: The point of this logic is to unset the attribute when it is present but invalid. Using `empty( $attr['decoding'] )` is not quite as accurate, for example if `$attr['decoding']` is not set, that condition will evaluate to `true`, thus unnecessarily also run the `in_array` check and the unnecessarily `unset` something that wasn't set in the first place.

So from my perspective this code looks good as is. FWIW, a similar condition is also used in the existing `loading` clause below.

Reply:
@felixarntz My apologies, it should have been `! empty( $attr['decoding'] )`.

Reply:
@felixarntz My suggestion was to change `array_key_exists() && ! in_array()` to `empty() || ! in_array()` - note the change from AND to OR as well.

`array_key_exists() && ! in_array()` will go through both conditions for `'0'`, `null`, `false`, `''`, `0`, `'0'`, `0.0`, `array()`, `new stdClass()`.

`empty() || ! in_array()` will skip the second condition for all of the above values. Yes, it will unnecessarily run `unset()` on a non-existent key, but `unset()` is a language construct, therefore more performant than `in_array()`.

The most performant option is `empty() || ! is_string() || ! in_array()`, which was my original suggestion in full.

Reply:
@adamsilverstein Some data on your questions:

Regarding how often `wp_get_attachment_image()` is called:
1. It's used by at least [5713 plugins](https://wpdirectory.net/search/01GQRPDFA00FW6RMHQ7YPW2S9H) and [3073 themes](https://wpdirectory.net/search/01GQRPFFCTJV9WV89GNS3H8MXW).
2. For each plugin/theme, this is done for every image it's called on. For themes, this is often called on every image in a gallery/product category.

Regarding the performance difference: https://3v4l.org/coWBm (times are multiplied by 1000 for easier reading)
- For invalid values, `array_key_exists()` is consistently slower.
- When `$attr['decoding']` isn't set, both `array_key_exists()` and `empty() + unset()` can take more than `0`. The time for `empty() + unset()` compared with `array_key_exists()` on its own is so close that it's interchangeable as you look at PHP 5.6 through 8.2.
- With both `empty()` and `unset()` being language constructs, these are two performant ways to skip `! in_array()` for values we already know are invalid.
- These times aren't significant for a single image, my suggestion is due to the cumulative impact.

Reply:
@costdev I gave this a quick performance benchmark test, calling `wp_get_attachment_image()` 1000 times in a single page load and measuring the total time that function takes using https://gist.github.com/felixarntz/de5c697a1a16c2b892634b70216eb6c7, using the median from 20 requests made.

Here are the results:
* `! array_key_exists() && ! in_array()`: 1301ms
* `empty() || ! in_array()`: 1290.44ms
* `empty() || ! is_string() || ! in_array()`: 1301.94ms

Based on the results, it is clear that the difference in performance is so small that continuing this discussion is probably not worth it (1000 times in a single page load are far from realistic). There's also typically a little (little, but some) variance in the data and because the difference is so small it needs to be taken with a grain of salt. But if that data shows anything beyond not much difference, it suggests that `empty() || ! in_array()` is a tiny bit faster. So I'm happy to go with that. 👍